Newsletters: add 274 (2023-10-25) (#1365)
bitcoinops/bitcoinops.github.io: _posts/en/newsletters/2023-10-25-newsletter.md

  implementations.  Recent versions include mitigations for the attack
  and upgrading to them is strongly recommended.  Only nodes being used
  to forward payments are affected; users who only use their channels to
  initiate and receive payments are not affected.  We've organized our
  description of this story into three separate news items: a
  description of the vulnerability (this item), a description of the
  mitigations deployed so far by various LN implementations, and a
  summary of mitigations and solutions proposed on the mailing list.

  _Replacement cycling_ is a technique that allows targeted removal of a
  transaction from node mempools.  To take a simple example, Mallory
  broadcasts a _precursor transaction_ that stays unconfirmed.  She then
  spends one of the outputs from the precursor in a _removable
  transaction_.  Later, she initiates a [replacement][] of precursor
  transaction, which triggers a removal of the removable transaction.

  When used by Mallory on transactions she exclusively controls, there's
  no significant problem with replacement cycling.  Except in cases where
  transactions get evicted from mempools for normal reasons, the BIP125

>   rules that Mallory's replacement must follew ensure that miners will

Reviewer: ```suggestion
  rules that Mallory's replacement must follow ensure that miners will
```

  (_MalloryB_).  MalloryB is supposed to either give Bob a _preimage_ that
  allows him to accept the forwarded payment from MalloryA, or MalloryB is
  supposed to cancel (revoke) the forwarded payment she received from Bob
  before a certain time.  Instead, MalloryB does nothing by the designated
  time and Bob is forced to close the channel and broadcast a transaction
  that spends the forwarded payment back to himself.  That spend should
  confirm promptly, allowing Bob to cancel (revoke) the spend he received
  from MalloryA, which returns everyone's balance to the amounts they were
  before the attempt to forward the payment (with the exception of any
  transaction fees paid to close and settle the Bob-MalloryB channel).

  Alternatively, when Bob closes the channel and attempts to spend the
  forwarded payment back to himself, MalloryB can replace his spend with a
  spend of her own containing the preimage.  If that transaction confirmed
  promptly, Bob would learn the preimage and be able to claim the
  forwarded payment from MalloryA, making Bob happy.

  However, if MalloryB replaces Bob's spend and a spend of her own that

>   contains the preimage but is also a removable tranaction, and she

Reviewer: ```suggestion
  contains the preimage but is also a removable transaction, and she
```

      transaction, that relay node will immediately be willing to accept
      Bob's spend again.  All Bob needs to do is re-broadcast his spend,
      which costs him nothing beyond the transaction fee he was already
      willing to pay.

      Before the private disclosure of replacement cycling, LN
      implementations only rebroadcasted their transactions infrequently
      (once per block or less).  There's normally a [privacy cost][topic
      transaction origin privacy] to broadcasting and rebroadcasting
      transactions---it might make it easier for third parties to
      associate Bob's LN activities with his IP address---although few
      public LN forwarding nodes currently try to hide this.  Now Core
      Lightning, Eclair, LDK, and LND will all rebroadcast more
      frequently.

      After each time Bob rebroadcasts, Mallory can use the same
      technique to replace his transaction again.  However, the BIP125
      replacement rules will require Mallory to pay additional
      transaction fees for each of her replacements and removals,
      meaning each rebroadcast by Bob lowers the profitability to

>       Mallory of a successful attack.

Reviewer: Since Bob’s transaction always has the same feerate and that feerate must be uncompetitive for the attack to be viable, is it really necessary for Mallory to increasingly pay more fees? If e.g. Mallory operates a high-volume wallet, she could use a UTXO to do the replacement cycle and then use that input towards an actual transaction at a competitive feerate.

Author: > Since Bob’s transaction always has the same feerate and that feerate must be uncompetitive for the attack to be viable

By "uncompetitive", do you mean that Bob must be paying too low of a feerate?  That's not actually true.  Bob could pay a feerate of 1,000 sat/vb and Mallory would still be able to perform a replacement cycle, which might still be worth it if earns her a 1,000,000 sat HTLC.

Was it something I wrote that gave you the impression that Bob's feerate must be uncompetitive?

> If e.g. Mallory operates a high-volume wallet, she could use a UTXO to do the replacement cycle and then use that input towards an actual transaction at a competitive feerate.

That's correct, but it's also where the deployed mitigations and proposed mitigations come in.  If Bob rebroadcasts 5 times a block, Mallory will need to have an operation where shes send about 5 organic transactions per block in order to have a marginal cost of zero.  That's fairly high volume, and it scales almost linearly depending on how many HTLCs Mallory is attacking (e.g., if she had 10 HTLCs in the channel with Bob, she would need to be at the level of typically sending about 50 organic transactions per block).  The additional proposed mitigation of having Bob pay higher fees as the CLTV expiry delta eliminates this advantage.  If Bob pays 10x the next-block fee near the end of CLTV expiry delta, then Mallory's replacements will need to pay an absolute fee greater than that amount, which she can only spread among additional outputs in organic transactions up to the 100,000 vbyte limit.  After that, she'll need to overpay fees.

In other words, I *think* the mitigations quickly make it so that only the highest of high-frequency spenders could have a minimal marginal attack cost, and that would only be possible if they only attacked a few HTLCs at once.  If they attacked more than about 100 HTLCs simultaneously (which is just 1 HTLC per block with a CLTV expiry delta of 144), I think anyone would be guaranteed to pay an above-zero marginal cost.

      This suggests a rough formula for the maximum amount of an HTLC
      that a node should accept.  If the cost is the attacker will need

>       to pay for each round of replacement-and-removal is _x_, the

Reviewer: ```suggestion
      that a node should accept.  If the cost that the attacker will need
      to pay for each round of replacement-and-removal is _x_, the
```

  [posted][riard cycle1] to the Bitcoin-Dev and Lightning-Dev mailing
  lists about a [responsibly disclosed][topic responsible disclosures]
  vulnerability affecting older versions of all LN implementations.
  Recent versions include mitigations for the attack and upgrading to
  them is strongly recommended.  Only nodes being used to forward
  payments are affected; users who only use their channels to initiate
  and receive payments are not affected.  We've organized our
  description of this story into three separate news items: a
  description of the vulnerability (this item), a description of the
  mitigations deployed so far by various LN implementations, and a
  summary of mitigations and solutions proposed on the mailing list.

  _Replacement cycling_ is a technique that allows targeted removal of a
  transaction from node mempools.  To take a simple example, Mallory
  broadcasts a _precursor transaction_ that stays unconfirmed.  She then
  spends one of the outputs from the precursor in a _removable
  transaction_.  Later, she initiates a [replacement][topic rbf] of the
  precursor transaction, which triggers a removal of the removable

>   transaction.

Author: @glozow 

Yeah, after writing this section and seeing a proposed fix that didn't seem to work, I [replied](https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2023-October/022067.html) to someone on list to also say that I didn't think the precursor is necessary.

I'll see if I can revise this section.  In the meantime, I [asked a question](https://bitcoin.stackexchange.com/questions/120200/what-are-the-advantages-to-attackers-of-different-styles-of-replacement-cycle-at) about the tradeoffs of the different approaches.

Thanks for pointing this out!

>   transaction.

Author: FYI, updated to use the simpler non-precursor scenario.   See https://github.com/bitcoinops/bitcoinops.github.io/pull/1365/commits/1f9d3eb61edacf2fabc83499a1c83a6b5b2f71ef (still working on other edits)

>   transaction.

Reviewer 2: Ah I didn't see that, thanks 👍

>   When used by Mallory on transactions she exclusively controls, there's

Reviewer 2: nitty nit
```suggestion
  When used by Mallory on transactions she exclusively controls, there is
```

        hot wallet connected to an always-on internet service---a type
        of deployment that is perpetually one vulnerability away from
        having all of its funds stolen.  Anyone evaluating the affect of
        replacement cycling on the risk profile of operating an LN
        forwarding node should consider it in the context of the risk
        that is already being tolerated.  Of course, it's worth
        searching for other ways to reduce that risk, as discussed in
        our next news item.

- **Proposed additional mitigations for replacement cycling:** as of
  this writing, there have been over 40 separate posts made to the
  Bitcoin-Dev and Lightning-Dev mailing list in response to the
  disclosure of the replacement cycling attack.  Suggested responses
  included the following:

    - **Incrementing fees towards scorched earth:** Antoine Riard's
      [paper][riard cycle paper] about the attack and mailing list posts
      by [Ziggie][ziggie cycle] and [Matt Morehouse][morehouse cycle]
      suggest that, instead of having the defender (e.g. Bob), just
      rebroadcast his refund spend, he start broadcasting conflicting

>       alternative spends that pay ever increasing feerates as the

Reviewer: Isn’t that just conjunctive mood?

      of it).  Theoretically, an attacker won't challenge a defender who
      they believe will follow a policy of mutually assured destruction,
      so Bob would never actually need to pay higher and higher
      feerates.  Whether that would be true in practice on the Bitcoin
      network is unproven.

    - **Automatic retrying of past transactions:** Corallo
      [suggested][corallo cycle1] that, "the only fix for this issue
      will be when miners keep a history of transactions they've seen
      and try them again after [...] an attack like this."  Bastien
      Teinturier [replied][teinturier cycle], "I agree with Matt though
      that more fundamental work most likely needs to happen at the
      bitcoin layer to allow L2 protocols to be more robust against that
      class of attacks."  Riard also [said][riard cycle3] something
      similar, "a sustainable fix can [only] happen at the base-layer,
      e.g adding a memory-intensive history of all seen transactions".

    - **Presigned fee bumps:** Peter Todd [argued][todd cycle1] that,
      "the correct way to do pre-signed transactions is to pre-sign

>       enough *different* transactions to cover all reasonable needs for

Author: Single asterisk and single underscore both produce italics in Markdown, so I'd prefer to stick with the verbatin copy/paste.

        The feerates used might be (in sats/vbyte): 1, 2, 4, 8, 16, 32,
        64, 128, 256, 512, 1024, which should cover anything for the
        foreseeable future.

        If MalloryB's preimage spend was presigned, the only replacement
        she could make would be to go from one feerate to a higher
        feerate.  She couldn't add new inputs to the preimage spend, and
        without that capability, she would be unable to initiate the
        removal part of a replacement cycle.

    - **OP_EXPIRE:** in a separate thread, but quoting from the
      replacement cycles thread, Peter Todd [proposed][todd expire1]
      several consensus changes to enable an `OP_EXPIRE` opcode will
      make a transaction invalid for inclusion after a specified block
      height if the transaction's script executes `OP_EXPIRE`.  This can
      be used to make the Mallory's preimage condition of an HTLC only
      usable before Bob's refund condition becomes spendable.  This
      prevents Mallory from being able to replace Bob's refund spend,
      making it impossible for Mallory to execute a replacement cycle
      attack.  `OP_EXPIRE` may also address some [transaction pinning

>       attacks][topic transaction pinning] against HTLCs.

Author: Good answer.  I upvoted it.  I didn't mention that concern here, in this very short summary, because Peter Todd took steps to address that concern in his proposal (to the same degree that coinbase transactions address that problem).

      enforced would be in addition to those vbytes).

    - _OP_SUCCESS changes would be beneficial:_ the [BIP342][]
      specification of [tapscript][topic tapscript] specifies several
      `OP_SUCCESSx` opcodes that make any script including them always
      succeed, allowing future soft forks to attach conditions to the
      opcodes (making them behave like regular opcodes).  However, that
      behavior makes it unsafe to use introspection with a covenant that
      allows including parts of an arbitrary script.  For example, Alice
      might want to create a covenant that allows her to spend her funds
      to an arbitrary address if she first spends her funds in a
      [vault][topic vaults] notification transaction and waits for some
      number of blocks to allow a freeze transaction to block the spend.
      However, if the arbitrary address includes an `OP_SUCCESSx`
      opcode, anyone will be able to steal her money.  Russell suggests
      two possible solutions to this problem.

  The research received some discussion and Russell indicated that he is
  working on a follow-up post related to introspection of output

>   amounts..

Reviewer 3: ```suggestion
  amounts.
```

---
title: 'Bitcoin Optech Newsletter #274'
permalink: /en/newsletters/2023/10/25/
name: 2023-10-25-newsletter
slug: 2023-10-25-newsletter
type: newsletter
layout: newsletter
lang: en
---
This week's newsletter describes the replacement cycling attack against
HTLCs used in LN and other systems, examines the mitigations deployed
for the attack, and summarizes several proposals for additional
mitigations.  Also described are a notable bug affecting a Bitcoin Core
RPC, research into covenants with minimal changes to Bitcoin Script, and
a proposed BIP for an `OP_CAT` opcode.  Also included is our regular
monthly section with summaries of popular questions and answers from the

> Bitcoin StackExchange.

Reviewer 4: ```suggestion
Bitcoin Stack Exchange.
```

{% assign bse = "https://bitcoin.stackexchange.com/a/" %}

## News

- **Replacement cycling vulnerability against HTLCs:** As briefly
  mentioned in [last week's newsletter][news274 cycle], Antoine Riard
  [posted][riard cycle1] to the Bitcoin-Dev and Lightning-Dev mailing
  lists about a [responsibly disclosed][topic responsible disclosures]
  vulnerability affecting all LN implementations.  Since the disclosure,
  implementations have been updated to include mitigations for the
  attack and we strongly recommend upgrading to the latest version of
  your preferred LN software.  Only nodes being used to forward payments
  are affected; users who only use their channels to initiate and
  receive payments are not affected.

  We've organized our description of this story into three separate news
  items: a description of the vulnerability (this item), a description
  of the mitigations deployed so far by various LN implementations, and
  a summary of additional mitigations and solutions proposed on the mailing list.

  As background, it's possible to use [transaction replacement][topic

>   rbf] to remove the one or more inputs of a multi-input transaction

Reviewer 5: ```suggestion
  rbf] to remove one or more inputs of a multi-input transaction
```

  from node mempools.  To take a simple example, one that [differs
  slightly]({{bse}}120200) from Riard's original description, Mallory
  broadcasts a transaction with two inputs, which spend outputs _A_ and
  _B_.  She then replaces that transaction with an alternative
  single-input version that only spends output _B_.  After that
  replacement, input _A_---and any data included in it---has been
  removed from any node mempools that processed the replacement.

>   Alhough it's not safe for a regular wallet to do that[^rbf-warning],

Reviewer 4: ```suggestion
  Although it's not safe for a regular wallet to do that[^rbf-warning],
```

  of the hash of the current UTXO set.  The hash did not commit to the
  height and coinbase information for each UTXO, information that is
  needed for enforcing the 100-block coinbase maturity rule and
  [BIP68][] relative timelocks.  All of that information is still in the
  database of a node that's been synced from scratch (all current
  Bitcoin Core nodes) and it's still used for enforcement, so this bug
  does not affect any known released software.  However, the
  experimental [assumeUTXO][topic assumeutxo] feature planned for
  the next major version of Bitcoin Core will allow users to share their
  UTXO databases with each other.  The incomplete commitment means that
  a modified database could have the same hash as a verified database,
  potentially opening a narrow window for an attack against assumeUTXO
  users.

  If you're aware of
  any software that uses the `hash_serialized_2` field, please notify
  its authors about the problem and encourage them to read Jahr's email
  about the changes being made for the next major version of Bitcoin
  Core to address the bug.

- **Research into generic covenants with minimal Script language changes:**

>   Rusty Russell [posted][russell scripts] to Bitcoin-Dev a link to some

Reviewer 4: ```suggestion
  Rusty Russell [posted][russell scripts] to Bitcoin-Dev mailing list a link to some
```

Reviewer 3: Took this suggestion with a slight addition

  purging 1.51sat/vb transactions while also indicating an estimated next block
  containing transactions with 1.49sat/vb. Glozow outlines the likely
  explanation as a full mempool resulting in the eviction of a transaction that
  bumped the node's mempool min feerate (by `-incrementalRelayFee`) but left
  some lower feerate transactions in the mempool that did not need to be
  evicted in order to stay within the max mempool size.

  She also mentions the asymmetry between [ancestor scoring][waiting for confirmation 2] for block
  template selection and descendant scoring for mempool eviction as another
  possible explanation and links to a [cluster mempool][topic cluster
  mempool]-related [issue][Bitcoin Core #27677] explaining the asymmetry and a
  potential new approach.

## Releases and release candidates

*New releases and release candidates for popular Bitcoin infrastructure
projects.  Please consider upgrading to new releases or helping to test
release candidates.*

- [Bitcoin Core 25.1][] is a maintenance release mainly containing bug
  fixes.  It is the current recommended version of Bitcoin Core.

- [Bitcoin Core 24.2][] is a maintenance relaese mainly containing bug

>   fixes.  It is recommend for anyone still using 24.0 or 24.1 who is

Reviewer 6: ```suggestion
  fixes.  It is recommended for anyone still using 24.0 or 24.1 who is
```

  unable or unwilling to upgrade to 25.1 at this time.

<!-- FIXME: bitschmidty to test whether the 26.0 link works before
publication.  If it does, please remove the "Verified test..." sentence.
In either case, please also delete this comment.  (Thanks, Mike!) -->

> - [Bitcoin Core 26.0][] is a release candidate for the next major

Reviewer 2: fyi @bitschmidty seems quite unlikely to have bins up before the newsletter is published.

Reviewer 6: Also:
```suggestion
<!-- FIXME: bitschmidty to test whether the 26.0 link works before
publication.  If it does, please remove the "Verified test..." sentence.
In either case, please also delete this comment.  (Thanks, Mike!) -->

- [Bitcoin Core 26.0rc1][] is a release candidate for the next major
```
